### PR TITLE
compute SCM information - update auto-assign algorithm

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: sonar-batch-protocol/src/main/java/org/sonar/batch/protocol/output/FileStructure.java [hunk 1]
```diff
@@ -27,7 +27,7 @@
 public class FileStructure {
 
   public static enum Domain {
-    ISSUES("issues-"), ISSUES_ON_DELETED("issues-deleted-"), COMPONENT("component-"), MEASURES("measures-");
+    ISSUES("issues-"), ISSUES_ON_DELETED("issues-deleted-"), COMPONENT("component-"), MEASURES("measures-"), SCMS("scms-");
```
Comment (henryju) [inline sonar-batch-protocol/src/main/java/org/sonar/batch/protocol/output/FileStructure.java]: This is only cosmetic but I don't really like the "S" on SCM*S*

File: sonar-batch-protocol/src/main/protobuf/batch_report.proto [hunk 2]
```diff
@@ -112,40 +112,58 @@ message Measures {
...
+  message Changeset {
+    optional string revision = 1;
+    optional string author = 2;
+    optional int64 date = 3;
+  }
+}
+
+message Scms {
+  optional int32 component_ref = 1;
+  repeated Scm scm = 2;
```
Comment (henryju) [inline sonar-batch-protocol/src/main/protobuf/batch_report.proto]: What is the purpose of this field? Please remove Scms message and move component_ref in Scm.

File: sonar-batch-protocol/src/main/java/org/sonar/batch/protocol/output/FileStructure.java (hunk 1 above)
Comment (teryk) [inline sonar-batch-protocol/src/main/java/org/sonar/batch/protocol/output/FileStructure.java]: yep, you're right. I wrongly thought SCM and SCMS were needed

**Comment by henryju** [comment]
LGTM

**Comment by teryk** [comment]
@julienlancelot please review https://github.com/SonarSource/sonarqube/commit/967a62a87c9fd3fbacb08da1f3698b025a724f5b

**Comment by teryk** [comment]
beware, I must not merge this pull-request into master before the feeding of scm information is done on batch side

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java [hunk 3]
```diff
@@ -19,35 +19,36 @@
...
-import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
 /**
- * Cache of the lines of the currently processed file. Only a single file
+ * Cache of the lines of the currently processed file. Only a <strong>single</strong> file
  * is kept in memory at a time. Moreover data is loaded on demand to avoid
  * useless db trips.
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java]: Update the comment as data is no more loaded from db but from the index

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java [hunk 4]
```diff
@@ -19,35 +19,36 @@
...
- * <p/>
+ * <br />
  * It assumes that db table FILE_SOURCES is up-to-date before using this
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java]: Same here

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java [hunk 5]
```diff
@@ -55,37 +56,41 @@
...
 
   /**
    * Marks the currently processed component
    */
-  void init(String fileUuid) {
+  void init(String fileUuid, @Nullable Integer fileReportRef) {
     loaded = false;
-    currentFileUuid = fileUuid;
-    authors.clear();
+    currentFileUuid = fileReportRef == null ? null : fileUuid;
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java]: Why currentFileUuid should be null when report is null ?

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java [hunk 6]
```diff
@@ -19,35 +19,36 @@
...
 /**
- * Cache of the lines of the currently processed file. Only a single file
+ * Cache of the lines of the currently processed file. Only a <strong>single</strong> file
  * is kept in memory at a time. Moreover data is loaded on demand to avoid
  * useless db trips.
- * <p/>
+ * <br />
  * It assumes that db table FILE_SOURCES is up-to-date before using this
  * cache.
  */
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java]: Please complete the comment to explain that data is first read from the report, then from the index

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java [hunk 7]
```diff
@@ -93,9 +98,42 @@ public String lineAuthor(@Nullable Integer lineId) {
    * Load only on demand, to avoid useless db requests on files without any new issues
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java]: Update comment

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java [hunk 8]
```diff
@@ -93,9 +98,42 @@ public String lineAuthor(@Nullable Integer lineId) {
...
    */
   private void loadIfNeeded() {
+    if (currentFileUuid == null || currentFileReportRef == null) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java]: Check null only on currentFileReportRef is enough

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java [hunk 9]
```diff
@@ -93,9 +98,42 @@ public String lineAuthor(@Nullable Integer lineId) {
...
+      throw new IllegalStateException(String.format("uuid (%s) and report reference (%d) must not be null to use the cache", currentFileUuid, currentFileReportRef));
+    }
     if (!loaded) {
-      dbClient.fileSourceDao().readDataStream(currentFileUuid, parserFunction);
-      loaded = true;
+      scm = reportReader.readComponentScm(currentFileReportRef);
+      loaded = scm != null;
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java]: Please extract following code is another method in order to understand that at this point of the code we're gonna load data from the index

**Comment by julienlancelot** [comment]
Apply my feedback, and LGTM !

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java (hunk 3 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java]: done

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java (hunk 5 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java]: done

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java (hunk 6 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java]: done

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java (hunk 7 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java]: done

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java (hunk 8 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java]: done

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java (hunk 9 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java]: done

**Comment by teryk** [comment]
@julienlancelot thank you to review https://github.com/SonarSource/sonarqube/commit/4f3f470000ab3e5da95c2132eb40106acab8b831 especially regarding the comment section

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java [hunk 10]
```diff
@@ -19,119 +19,142 @@
...
 
 /**
- * Cache of the lines of the currently processed file. Only a single file
+ * Cache of the lines of the currently processed file. Only a <strong>single</strong> file
  * is kept in memory at a time. Moreover data is loaded on demand to avoid
- * useless db trips.
- * <p/>
- * It assumes that db table FILE_SOURCES is up-to-date before using this
- * cache.
+ * useless db trips.<br />
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java]: db ? :)

File: server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java [hunk 11]
```diff
@@ -19,119 +19,142 @@
...
+ * It relies on:
+ * <ul>
+ *   <li>the SCM information sent in the report for modified files</li>
+ *   <li>the source line index for non-modified files</li>
+ * </ul>
+ *
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/issue/SourceLinesCache.java]: :+1:

**Comment by julienlancelot** [comment]
One last comment to update and LGTM

**Comment by teryk** [comment]
@henryju you can feed SCM information when you have time. PR affected to you.

**Pull request merged**